### PR TITLE
flaky tests: consistent logging

### PR DESCRIPTION
**Pull request by a-palchikov**
This PR simplifies the logger set up for tests: the logger is setup consistently with `DebugLevel` and outputting to `os.Stderr`, and formatter with timestamps since timing skews might be highly relevant for troubleshooting time-sensitive code.
I'm also bumping logrus to v1.6.0 since I found that our fork has data races and the fork is pretty far behind that release.
Running at `DebugLevel` also discovered a legitimate bug in roles code.

Requires https://github.com/gravitational/teleport.e/pull/183.

### CODE REVIEW

File: lib/events/auditlog.go [hunk 1]
```diff
@@ -109,7 +109,7 @@ func init() {
 // sessions. It implements IAuditLog
 type AuditLog struct {
 	sync.Mutex
-	*log.Entry
+	log.FieldLogger
```
Comment (a-palchikov) [inline lib/events/auditlog.go:112]: `logrus.Entry` in v1.6.0 embeds a `context.Context` which causes disambiguation errors here and below.

File: lib/services/role.go [hunk 2]
```diff
@@ -1501,7 +1501,7 @@ func MatchLabels(selector Labels, target map[string]string) (bool, string, error
 // RoleNames returns a slice with role names. Removes runtime roles like
 // the default implicit role.
 func (set RoleSet) RoleNames() []string {
-	out := make([]string, len(set)-1)
```
Comment (a-palchikov) [inline lib/services/role.go:1504]: this was misbehaving for `len(RoleSet) <= 1`

File: go.mod [hunk 3]
```diff
@@ -106,5 +105,4 @@ require (
 replace (
 	github.com/coreos/go-oidc => github.com/gravitational/go-oidc v0.0.3
 	github.com/iovisor/gobpf => github.com/gravitational/gobpf v0.0.1
-	github.com/sirupsen/logrus => github.com/gravitational/logrus v0.10.1-0.20171120195323-8ab1e1b91d5f
```
Comment (awly) [inline go.mod:109]: Do you know why we made a fork?
Hope we're not losing some internal patch that teleport depends on.

File: lib/events/auditlog.go (hunk 1 above)
Comment (awly) [reply lib/events/auditlog.go:112]: Could you refactor this to not embed the logger (name this field `log`)?
Otherwise we pollute the type's API for a few saved characters.

File: lib/kube/proxy/forwarder.go [hunk 4]
```diff
@@ -216,7 +216,7 @@ func NewForwarder(cfg ForwarderConfig) (*Forwarder, error) {
 // however some requests like exec sessions it intercepts and records.
 type Forwarder struct {
 	sync.Mutex
-	*log.Entry
+	log.FieldLogger
```
Comment (awly) [inline lib/kube/proxy/forwarder.go:219]: same here, please name the field and refactor all call sites.

File: lib/reversetunnel/srv.go [hunk 5]
```diff
@@ -104,7 +104,7 @@ type server struct {
 	// ctx is a context used for signalling and broadcast
 	ctx context.Context
 
-	*log.Entry
+	log.FieldLogger
```
Comment (awly) [inline lib/reversetunnel/srv.go:107]: same here - name the field

File: lib/utils/cli_test.go [hunk 6]
```diff
@@ -36,31 +36,34 @@ var _ = check.Suite(&CLISuite{})
 func (s *CLISuite) SetUpSuite(c *check.C) {
 	InitLoggerForTests()
 }
-func (s *CLISuite) TearDownSuite(c *check.C) {}
-func (s *CLISuite) SetUpTest(c *check.C)     {}
-func (s *CLISuite) TearDownTest(c *check.C)  {}
 
 func (s *CLISuite) TestUserMessageFromError(c *check.C) {
+	// FIXME: enable after https://drone.gravitational.io/gravitational/teleport/3517 is merged
+	c.Skip("TODO")
```
Comment (awly) [inline lib/utils/cli_test.go:42]: Leaving a comment to resolve after this is fixed

**Review by awly — COMMENTED** [review]
Great cleanup, thanks!

File: lib/reversetunnel/srv_test.go [hunk 7]
```diff
@@ -38,7 +39,7 @@ func TestServerKeyAuth(t *testing.T) {
 	require.NoError(t, err)
 
 	s := &server{
-		Entry: testlog.FailureOnly(t),
+		log: utils.NewLoggerForTests(logrus.New()),
```
Comment (awly) [inline lib/reversetunnel/srv_test.go:42]: `testlog.FailureOnly` has a different behavior - it buffers all logs and only prints them when `t` fails.
please revert this change

File: go.mod (hunk 3 above)
Comment (awly) [reply go.mod:109]: Awesome, hopefully we can archive our fork after this merges!

**Review by awly — APPROVED** [review]

File: lib/reversetunnel/srv_test.go (hunk 7 above)
Comment (a-palchikov) [reply lib/reversetunnel/srv_test.go:42]: I believe testing has this behavior by default - it buffers all output (sent via `t.Log(f)` or  `os.StdXXX`) and only outputs it if the test fails regardless of the `-v` (verbose) option. The verbose option is only relevant if one wants to see logs regardless of the outcome.
Do you mean that this should suppress log output regardless of verbosity and only output it on failure?

I initially implemented `InitLoggerForTests` to accept the value of `*testing.T` and route the logs using its `Log(f)` APIs but it became a bit of concern to use with a global logrus logger since it should not outlive a test so I abandoned this approach in favor of writing to `os.Stderr` as this is also buffered by the testing package with the behavior described above.

Comment (awly) [reply lib/reversetunnel/srv_test.go:42]: the default behavior in `testing` doesn't know which test outputs which logs.
so if you have a package with 100 tests and 1 of them fails, you get logs for all 100 tests intermixed with test result output.

and yes, the reason I wrote `testlog.FailureOnly` is because it's scoped to a single `testing.T` and not globally

Comment (a-palchikov) [reply lib/reversetunnel/srv_test.go:42]: The whole point of changes in this PR is exactly this behavior - hence I made this change too. Like you said, it is not going to be helpful if all of the tests' logs are output upon failure (which will happen if this is merged), so I might need to reconsider the approach I've taken with this.

The goal is getting the logs only for the failed test - no doubt about that but also that it should be universal - i.e. all tests should use the FailureOnly logger otherwise it's a hit or miss if used selectively. I'll think about a way to get this goal instead.

Comment (awly) [reply lib/reversetunnel/srv_test.go:42]: One other option is to set `logrus.Logger.Out` to a wrapper around `t.Log/c.Log`
It works for both `gocheck` and `testing` tests.

The challenge is making sure that none of our code uses the global logrus logger, but that can be done incrementally afterwards.

File: lib/reversetunnel/srv_test.go [hunk 8]
```diff
@@ -38,7 +38,7 @@ func TestServerKeyAuth(t *testing.T) {
 	require.NoError(t, err)
 
 	s := &server{
-		Entry: testlog.FailureOnly(t),
+		log: utils.NewLoggerForTests(),
```
Comment (awly) [inline lib/reversetunnel/srv_test.go:41]: let's revert this change, `testlog.FailureOnly` seems appropriate for non-gocheck tests

File: lib/service/service.go [hunk 9]
```diff
@@ -83,9 +84,7 @@ import (
 	"github.com/sirupsen/logrus"
 )
 
-var log = logrus.WithFields(logrus.Fields{
-	trace.Component: teleport.ComponentProcess,
-})
+var log = logrus.WithField(trace.Component, teleport.ComponentProcess)
```
Comment (awly) [inline lib/service/service.go:87]: looks like there are still a few references to this package-global logger instead of `process.log` below.
remove this var and fix any compilation errors below.

File: lib/srv/app/server.go [hunk 10]
```diff
@@ -440,6 +441,7 @@ func (s *Server) newHTTPServer() *http.Server {
 	return &http.Server{
 		Handler:           authMiddleware,
 		ReadHeaderTimeout: defaults.DefaultDialTimeout,
+		ErrorLog:          stdlog.New(utils.NewStdlogAdaptor(s.log.Error), teleport.ComponentApp, stdlog.LstdFlags),
```
Comment (awly) [inline lib/srv/app/server.go:444]: this looks like a common pattern.
how about pulling it into the utils package, something like
```
func NewStdlog(logger leveledOutputFunc, component string) *stdlog.Logger
```

File: lib/utils/cli.go [hunk 11]
```diff
@@ -207,6 +227,41 @@ func EscapeControl(s string) string {
 	return s
 }
 
+// Write writes the specifies buffer p to the underlying logger at INFO level.
```
Comment (awly) [inline lib/utils/cli.go:230]:
```suggestion
// Write writes the specifies buffer p to the underlying logger.
```

File: lib/utils/testlog/log.go [hunk 12]
```diff
@@ -20,19 +20,19 @@ package testlog
...
-	"github.com/sirupsen/logrus"
+
+	"gopkg.in/check.v1"
 )
 
 // FailureOnly returns a logger that only prints the logs to STDERR when the
 // test fails.
-func FailureOnly(t *testing.T) *logrus.Entry {
-	// Collect all the output in buf.
+func FailureOnly(t testingInterface) utils.Logger {
```
Comment (awly) [inline lib/utils/testlog/log.go:31]: the argument type should be exported

File: lib/utils/testlog/log.go [hunk 13]
```diff
@@ -42,5 +42,64 @@ func FailureOnly(t *testing.T) *logrus.Entry {
...
+func NewCheckTestWrapper(c *check.C) *TestWrapper {
+	w := &TestWrapper{
+		c: c,
+	}
+	w.Log = FailureOnly(w)
+	return w
+}
+
+// Cleanup registers the specified handler f to be run
+// after the hest gas completed
```
Comment (awly) [inline lib/utils/testlog/log.go:63]:
```suggestion
// after the test has completed
```

**Review by awly — APPROVED** [review]

File: integration/integration_test.go [hunk 14]
```diff
@@ -1115,13 +1146,18 @@ func (s *IntSuite) runDisconnectTest(c *check.C, tc disconnectTestCase) {
 
 	select {
 	case <-time.After(tc.disconnectTimeout + time.Second):
-		c.Fatalf("timeout waiting for session to exit: %+v", tc)
+		pprof.Lookup("goroutine").WriteTo(os.Stderr, 1)
```
[the remark this answers is not in the document]
Comment (a-palchikov) [reply integration/integration_test.go:1149]: Usually, with the (longer) timeout that subsequently fails the test, it is helpful to know where the process is stuck, hence this is not trying to dump the current goroutine's stack but the processes state (e.g. akin to `go test` when it terminates the test process after a timeout).
So I would not expect `runtime.Stack(buf, true)` to be more appropriate here as there is no need to buffer the result - it is fine to stream it directly to stderr.

File: lib/reversetunnel/srv_test.go (hunk 8 above)
Comment (a-palchikov) [reply lib/reversetunnel/srv_test.go:41]: Yes, sorry - missed the spot.

File: integration/integration_test.go (hunk 14 above)
[the remark this answers is not in the document]
Comment (awly) [reply integration/integration_test.go:1149]: makes sense.
I assumed that using `pprof` here required special flags to `go test` to collect the profile, but it seems to work out-of-the-box.
this is much cleaner than `runtime.Stack` :+1:

File: lib/utils/testlog/log.go (hunk 13 above)
Comment (a-palchikov) [reply lib/utils/testlog/log.go]: Addressed in https://github.com/gravitational/teleport/pull/4849/commits/756c17b66db83fc5637415d0231fd0324cd72519

File: integration/app_integration_test.go [hunk 15]
```diff
@@ -238,6 +239,8 @@ func setup(t *testing.T) *pack {
 	tr := utils.NewTracer(utils.ThisFunction()).Start()
 	defer tr.Stop()
 
+	log := testlog.FailureOnly(t)
```
Comment (r0mant) [inline integration/app_integration_test.go:242]: Does this mean integration tests are not going to output info/warn messages anymore, or is there a way to toggle them on? Just asking cause as I was working on an integration tests for db access, I used info outputs quite a bit to troubleshoot issues.

File: go.mod (hunk 3 above)
Comment (r0mant) [reply go.mod:109]: @a-palchikov With this change will `WithError(err)` finally start behaving properly by printing the error with a stack trace, or would we need to additionally bump logrus to get it working?

File: integration/integration_test.go [hunk 16]
```diff
@@ -119,7 +121,7 @@ func (s *IntSuite) SetUpSuite(c *check.C) {
 
 	// close & re-open stdin because 'go test' runs with os.stdin connected to /dev/null
 	stdin, err := os.Open("/dev/tty")
-	if err != nil {
+	if err == nil {
```
Comment (r0mant) [inline integration/integration_test.go:124]: I'm curious if this should just assert `c.Assert(err, check.IsNil)`, or is it an expected scenario, do you know? Kinda weird that tests pass with both this condition and (presumably) erroneous previous condition.

**Review by r0mant — APPROVED** [review]

File: integration/app_integration_test.go (hunk 15 above)
Comment (a-palchikov) [reply integration/app_integration_test.go:242]: No, the test output will be streamed as expected in verbose mode.

File: integration/integration_test.go (hunk 16 above)
Comment (a-palchikov) [reply integration/integration_test.go:124]: This does look suspicious. The original implementation did just that and was changed [here](https://github.com/gravitational/teleport/commit/e4023519d0bea20e187e1666f933fe9ff72e8b7a).

Now, aside from that. the real question is why stdin is required in an integration test? As long as we can input the data from some file, it should be a good replacement for the code that'd require user input otherwise, so it's sort of confusing. I'll look at this in a later PR.

Regarding the logrus update - no, `WithError` will not wrap errors as you imagine since this is gravitational/trace functionality (namely, the TextFormatter implemented therein) which is only available in `trace` 1.1.10 and on. I'll bump it separately.
